shared-module/audiomp3/MP3Decoder.c

```diff
@@ -7,93 +7,188 @@
 
 #include "shared-bindings/audiomp3/MP3Decoder.h"
 
+#include <math.h>
 #include <stdint.h>
 #include <string.h>
-#include <math.h>
+#include <sys/types.h>
+#include <unistd.h>
 
 #include "py/mperrno.h"
 #include "py/runtime.h"
+#include "py/stream.h"
 
 #include "shared-module/audiomp3/MP3Decoder.h"
 #include "supervisor/background_callback.h"
 #include "lib/mp3/src/mp3common.h"
 
 #define MAX_BUFFER_LEN (MAX_NSAMP * MAX_NGRAN * MAX_NCHAN * sizeof(int16_t))
 
+#define DO_DEBUG (0)
+
+#if defined(MICROPY_UNIX_COVERAGE)
+#define background_callback_prevent() ((void)0)
+#define background_callback_allow() ((void)0)
+#define background_callback_add(buf, fn, arg) ((fn)((arg)))
+#endif
+
+static bool stream_readable(void *stream) {
+    int errcode = 0;
+    mp_obj_base_t *o = MP_OBJ_TO_PTR(stream);
+    const mp_stream_p_t *stream_p = MP_OBJ_TYPE_GET_SLOT(o->type, protocol);
+    if (!stream_p->ioctl) {
+        return true;
+    }
+
+    mp_int_t ret = stream_p->ioctl(stream, MP_STREAM_POLL, MP_STREAM_POLL_RD | MP_STREAM_POLL_ERR | MP_STREAM_POLL_HUP, &errcode);
+    if (DO_DEBUG) {
+        mp_printf(&mp_plat_print, "stream_readable ioctl() -> %d [errcode=%d]\n", ret, errcode);
+    }
+    return ret != 0;
+}
+
+// (near copy of mp_stream_posix_read, but with changes)
+// (circuitpy doesn't enable posix stream routines anyway)
```

Comment:
What changes? Would it be better to enable the shared posix routines?

Reply:
MICROPY_STREAMS_POSIX_API enables a number of routines, but we wouldn't be using them all. Probably the linker would avoid pulling them in though?

the posix wrappers also uses the global `errno` value, which I'd rather avoid (but it does make sense as a POSIX-style wrapper).

`mp_stream_rw` or calls that end up inside it are not suitable (*there is no `mp_stream_read` just `mp_stream_write` for some reason), because we need to take care to check for readability before *EACH* read call, but `mp_stream_rw` calls the read routine in a loop.

Reply:
I updated the comments to be clearer about what is different and why.

Reply:
Ok, thanks! I don't like `errno` either so I'm happy to avoid it.